### PR TITLE
Updates to sys.monitoring to get debugpy tests to pass

### PR DESCRIPTION
**Pull request by rchiodo**
I submitted these changes (and plus all of the other sys.monitoring changes) to debugpy here:
https://github.com/microsoft/debugpy/pull/1680

These changes get all of the tests in debugpy to pass with 3.12.

### CODE REVIEW

File: _pydevd_bundle/pydevd_safe_repr.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ class SafeRepr(object):
     # most level, and truncated to maxstring_inner characters inside
     # collections.
     maxstring_outer = 2**16
-    maxstring_inner = 30
+    maxstring_inner = 128
```
Comment (rchiodo) [inline _pydevd_bundle/pydevd_safe_repr.py:23]: This isn't entirely necessary but was a change that debugpy made a while ago.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 2]
```diff
@@ -643,6 +661,7 @@ def _enable_return_tracing(code):
 # ENDIF
 # fmt: on
     # print('enable return tracing', code)
+    _ensure_monitoring()
```
Comment (rchiodo) [inline _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:664]: This function was being called prior to the debugger being setup as the sys.monitor

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 3]
```diff
@@ -168,51 +169,67 @@ def _get_bootstrap_frame(depth: int) -> Tuple[Optional[FrameType], bool]:
...
+            return False
 
-        return f_unhandled
+    return True
 
+# fmt: off
+# IFDEF CYTHON
+# cdef _is_last_user_frame(frame: FrameType):
+# ELSE
+def _is_last_user_frame(frame: FrameType) -> bool:
```
Comment (rchiodo) [inline _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:216]: I changed the logic for get_unhandled_frame to just check if the passed in frame was the last user frame. When using Cython the old logic wasn't working because getFrame wasn't returning any of the pydevd frames and the logic was off.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 4]
```diff
@@ -897,15 +922,23 @@ def _raise_event(code, instruction, exc):
...
+    _thread_local_info._user_uncaught_exc_info = _user_uncaught_exc_info
+
     # print('!!!! should_stop (in raise)', should_stop)
     if should_stop:
         handle_exception(py_db, thread_info.thread, frame, arg, EXCEPTION_TYPE_HANDLED)
-        return
+
+    # Once we leave the raise event, we are no longer in the state of 'just_raised', so 
+    # indicate that this traceback is for an exception in the unwinding state
+    flag_as_unwinding(exc.__traceback__)
```
Comment (rchiodo) [inline _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:941]: The django plugin was messing up when trying to determine if an exception was just raised or not. This special flag makes that work.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 5]
```diff
@@ -1333,6 +1366,10 @@ def _jump_event(code, from_offset, to_offset):
 
     # We know the frame depth.
     frame = _getframe(1)
+
+    # Disable the next line event as we're jumping to a line. The line event will be redundant.
```
Comment (rchiodo) [inline _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1370]: One of the debugpy tests found that breakpoints were hitting too often. This prevents double events for the same line when jumping

File: pydev_ipython/inputhook.py [hunk 6]
```diff
@@ -176,9 +177,11 @@ def disable_wx(self):
         self.clear_inputhook()
 
     def enable_qt(self, app=None):
-        from pydev_ipython.qt_for_kernel import QT_API, QT_API_PYQT5
+        from pydev_ipython.qt_for_kernel import QT_API, QT_API_PYQT5, QT_API_PYQT6
```
Comment (rchiodo) [inline pydev_ipython/inputhook.py:180]: Debugpy had added support for qt6 here. Thought it would be good to include it.

File: pydevd.py [hunk 7]
```diff
@@ -1077,7 +1077,18 @@ def get_file_type(self, frame, abs_real_path_and_basename=None, _cache_file_type
             if abs_real_path_and_basename[0] == "<string>":
                 # Consider it an untraceable file unless there's no back frame (ignoring
                 # internal files and runpy.py).
+                if frame.f_back is None:
+                    _cache_file_type[cache_key] = None
+                    return None
+                
+                back_basename = pydevd_file_utils.basename(frame.f_back.f_code.co_filename)
```
Comment (rchiodo) [inline pydevd.py:1084]: get_file_type was failing in sys.monitoring because a bunch of <string> frames were coming from the code in pydevd.

**Review by fabioz — CHANGES_REQUESTED** [review]
As a note, especially on cases where subtle things were changed in the monitoring, the lack of tests along in the PR which check the actual use-case is the major blocker for the PR integration... Do you think you can add those?

As a note, I took some time to make the `main` branch green (some things were failed because of updates out of the debugger), so, you can add more tests and verify that your changes don't break the current test suite (`pydevd` has a much higher test coverage on its internals than `debugpy`, so, I'm really interested on what tests slipped from the coverage that still did make it up to `debugpy`).

File: pydev_ipython/version.py [hunk 8]
```diff
@@ -1,50 +1,36 @@
...
-    def __ge__(self, v):
-        this_parts = self._version_parts
-        other_parts = v._version_parts
+# -----------------------------------------------------------------------------
+# Imports
+# -----------------------------------------------------------------------------
 
-        while len(this_parts) < len(other_parts):
-            this_parts = this_parts + (0,)
+from distutils.version import LooseVersion
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply pydev_ipython/version.py:19]: I just copied this over from debugpy (wasn't part of the change I made). Hopefully @int19h knows the reason behind this change? I can just revert it in this PR for now.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 3 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:216]: The tests that were failing were these:

https://github.com/microsoft/debugpy/blob/25955a05d87a6e3d7694f7479c1c17ded0f9af90/tests/debugpy/test_exception.py#L160

The unhandled frame logic (when Cython was active) would fail to catch that something was unhandled.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 4 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:941]: I'll create a test that matches the one in debugpy. I believe this is the one that was failing:
https://github.com/microsoft/debugpy/blob/25955a05d87a6e3d7694f7479c1c17ded0f9af90/tests/debugpy/test_django.py#L138

It was in the django plugin where it was trying to determine if a template exception was being thrown I believe.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 5 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1370]: This was the test that was failing in debugpy. On this specific line:

https://github.com/microsoft/debugpy/blob/25955a05d87a6e3d7694f7479c1c17ded0f9af90/tests/debugpy/test_breakpoints.py#L152

The breakpoints would fire twice as much as expected.

File: pydevd.py (hunk 7 above)
Comment (rchiodo) [reply pydevd.py:1084]: I definitely ran the build tools (it's how I updated the cython files).

I should remove the logging though. Forgot about that. I need to debug this again to see what the exact problem was.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
     RETURN_VALUES_DICT,
     PYTHON_SUSPEND,
 )
+from _pydevd_bundle.pydevd_frame_utils import short_tb, flag_as_unwinding, short_frame
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:25]: Oh yes, forgot that. Will do.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 10]
```diff
@@ -810,6 +830,9 @@ def _unwind_event(code, instruction, exc):
         if thread_info is None:
             return
 
+    frame = _getframe(1)
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:833]: Sorry it was for debugging purposes. I can move it back.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 11]
```diff
@@ -882,6 +904,9 @@ def _raise_event(code, instruction, exc):
...
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:907]: Same here. For debugging purposes. I was printing the frames to try and figure out why tests were failing

File: pydevd.py [hunk 12]
```diff
@@ -1094,6 +1105,8 @@ def get_file_type(self, frame, abs_real_path_and_basename=None, _cache_file_type
                         # to show it in the stack.
                         _cache_file_type[cache_key] = LIB_FILE
                         return LIB_FILE
+                    
+                    back_frames += " -> %s" % (pydevd_file_utils.basename(f.f_code.co_filename))
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply pydevd.py:1109]: Yes, will do.

File: pydevd_attach_to_process/windows/compile_windows.bat [hunk 13]
```diff
@@ -1,3 +1,7 @@
+:: This script compiles the attach and inject DLLs for x86 and x64 architectures.
+:: It runs as part of the PR checks in the debugpy repo, AND the artifacts are uploaded to github,
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply pydevd_attach_to_process/windows/compile_windows.bat:2]: Will do.

File: pydevd.py [hunk 14]
```diff
@@ -2386,6 +2399,7 @@ def set_trace_for_frame_and_parents(self, thread_ident: Optional[int], frame, **
         while frame is not None:
             if not isinstance(frame, FrameType):
                 # This is the case for django/jinja frames.
+                pydev_log.debug("Not a frame: %s", frame)
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply pydevd.py:2402]: Will do.

File: _pydevd_bundle/pydevd_cython.pyx [hunk 15]
```diff
@@ -1396,7 +1396,7 @@ def should_stop_on_exception(py_db, PyDBAdditionalThreadInfo info, frame, thread
             exception_breakpoint = None
             try:
                 if py_db.plugin is not None:
-                    result = py_db.plugin.exception_break(py_db, frame, thread, arg)
+                    result = py_db.plugin.exception_break(py_db, frame, thread, arg, is_unwind)
```
Comment (rchiodo) [inline _pydevd_bundle/pydevd_cython.pyx:1399]: This is a different way to solve the django issue. It checks to make sure we're not during an unwind. In 3.11 this isn't a problem because the cached exception prevents the unwind event from firing.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 5 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1365]: I added another test for this and fixed the special cases in the tests for 3_12 that this also fixes

File: pydevd.py (hunk 7 above)
Comment (rchiodo) [reply pydevd.py:1084]: Sorry didn't get to this one yet. Need to debug still.

File: pydevd_attach_to_process/windows/attach.cpp [hunk 16]
```diff
@@ -108,7 +107,7 @@ struct InitializeThreadingInfo {
     PyImport_ImportModule* pyImportMod;
     PyEval_Lock* initThreads;
 
-    std::mutex mutex;
+    CRITICAL_SECTION cs;
```
Comment (rchiodo) [inline pydevd_attach_to_process/windows/attach.cpp:110]: This fixes a problem with attach in 3.12. The mutex was throwing an access violation. This worked more consistently.

File: tests_python/debugger_unittest.py [hunk 17]
```diff
@@ -647,15 +647,20 @@ def run_process(self, args, writer):
             except:
                 traceback.print_exc()
             finish[0] = True
+            print("Log on success: " + self.get_log_contents())
```
Comment (rchiodo) [inline tests_python/debugger_unittest.py:650]: Used this for test comparisons. I'll comment it out.

File: tests_python/test_debugger.py [hunk 18]
```diff
@@ -3109,16 +3109,6 @@ def test_remote_debugger_threads(case_setup_remote):
         writer.write_run_thread(hit_in_main.thread_id)
         writer.write_run_thread(hit_in_thread1.thread_id)
         writer.write_run_thread(hit_in_thread2.thread_id)
-
-        if TODO_PY312:
```
Comment (rchiodo) [inline tests_python/test_debugger.py:3113]: This code was only necessary because of the extra breakpoints firing with the line/jump events. With my fix in place this is no longer needed.

File: tests_python/test_debugger_json.py [hunk 19]
```diff
@@ -729,6 +729,30 @@ def accept_message(msg):
 
         writer.finished_ok = True
 
+def test_case_json_hit_condition_error_count(case_setup_dap):
```
Comment (rchiodo) [inline tests_python/test_debugger_json.py:732]: THis was the test case that debugpy caught the line/jump extra events with

File: tests_python/test_debugger_json.py [hunk 20]
```diff
@@ -4640,7 +4666,7 @@ def test_case_django_no_attribute_exception_breakpoint(case_setup_django_dap, jm
                     "protected": "inline",
                 },
             )
-            json_facade.write_set_exception_breakpoints(["raised"])
+            json_facade.write_set_exception_breakpoints(["raised", "uncaught"])
```
Comment (rchiodo) [inline tests_python/test_debugger_json.py:4669]: This reproduces the django exception that I added the 'is_unwind' parameter for.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 4 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py]: Updated your test in test_debugger_json that validates this change. Without it the test fails.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 3 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Sorry still need to do this too. Haven't gotten to debugging this with the old code yet.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 4 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py]: I changed the logic though to pass through whether or not we're 'unwinding' instead of holding onto the traceback

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 3 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Going back to the old way, all of these tests fail in debugpy:

```
FAILED tests/debugpy/test_exception.py::test_systemexit[0-zero-uncaught-raised-attach_connect(cli)-program]
... (38 in total)
```

I'm looking into how to create a similar test in PyDev.Debugger.

Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Hmm, there's already a bunch of `test_case_sys_exit_unhandled_exception` type cases. Not sure why they have different results.

Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: I added some tests that simulate what happens in debugpy failures but they don't fail with the old code. It seems the old code only fails on 'attach'. (which I haven't figured out how to do in pydevd yet)

Would attach change how sys.getframe worked?

Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Okay figured out how to write an attach test. Still working on how this is failing with the old code

Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Still doesn't fail with pydevd. Blech. I'll have to debug the difference in debugpy then.

Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Okay in debugpy these attach tests throw more unhandled exceptions, maybe because jmc isn't being applied?

I can see a whole bunch more exceptions:

```
0.00s - Ignore exception <class 'AttributeError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (__str__)
0.00s - Ignore exception <class 'AttributeError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (drive)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (open)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (open)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (read_text)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (read_text)
0.00s - File not traced (not in project): C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\site-packages\importlib_metadata\__init__.py
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\site-packages\importlib_metadata\__init__.py -- (read_text)
... // There's a bunch more
```

This results in this code here getting hit earlier in `_get_unhandled_exception_frame`

```python

        if f_unhandled is not None:
            _thread_local_info.f_unhandled = f_unhandled
            return _thread_local_info.f_unhandled
```

But there's nothing that ever unsets this `f_unhandled` for the current thread. So it stays there. 

This prevents future exceptions from being marked as unhandled because this exception frame is reused. 

Removing this caching fixes the problem.

Trying to figure out a way to reproduce this in a test. Maybe have the test throw more exceptions in a different frame before the sys.exit.

Comment (fabioz) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: I think that the problem may be that it's just caching one when possibly other entries arise?

The logic here is that the cache is caching the first frame where things should be considered unhandled

So, in a stack such as

```
runpy.run
user code
```
it should mark `runpy.run` as the top thing, but there are cases when multiple top frames could be considered

i.e.:

```
threading.bootstrap1
threading.bootstrap2
user code
```

in this case all of those could be considered

So, the logic could be changed to keep on going up the stack and instead of keeping just 1 keep all that match to see if it works (so, instead of `_get_unhandled_exception_frame` it'd be `_get_unhandled_exception_frames`).

If that doesn't work, please just comment the cache out instead of doing the refactor and then this can be checked after the PR is merged.

File: _pydevd_bundle/pydevd_cython.pyx [hunk 21]
```diff
@@ -1416,7 +1416,7 @@ def should_stop_on_exception(py_db, PyDBAdditionalThreadInfo info, frame, thread
                     pass
 
                 else:
-                    was_just_raised = trace.tb_next is None
+                    was_just_raised = just_raised(trace)
```
Comment (fabioz) [inline _pydevd_bundle/pydevd_cython.pyx:1419]: Please inline this function (the difference is that it'll check if trace itself is None, just do that inline here.

If this wasn't a debugger, I'd prefer the function call, but for the debugger, additional function calls should be avoided for performance reasons -- especially here, this is all in cython and `just_raised` isn't in a cythonized module.

File: _pydevd_bundle/pydevd_process_net_command_json.py [hunk 22]
```diff
@@ -345,6 +345,7 @@ def _resolve_remote_root(self, local_root, remote_root):
         return remote_root
 
     def _set_debug_options(self, py_db, args, start_reason):
+        pydev_log.debug("RCHIODO -- setting debug options")
```
Comment (fabioz) [inline _pydevd_bundle/pydevd_process_net_command_json.py:348]: I guess this can be removed now.

File: pydev_ipython/inputhookqt4.py [hunk 23]
```diff
@@ -19,7 +19,8 @@
 import os
 import signal
 
-from _pydev_bundle._pydev_saved_modules import threading
+import threading
```
Comment (fabioz) [inline pydev_ipython/inputhookqt4.py:22]: Why was this change needed? Usually it's preferred to use the pydevd import as it'll get the original module when using gevent.

File: pydev_ipython/qt_loaders.py [hunk 24]
```diff
@@ -109,23 +118,24 @@ def has_binding(api):
...
         QT_API_PYQT: "PyQt4",
         QT_API_PYQTv1: "PyQt4",
         QT_API_PYQT_DEFAULT: "PyQt4",
         QT_API_PYQT5: "PyQt5",
+        QT_API_PYQT6: "PyQt6",
     }
     module_name = module_name[api]
 
-    import importlib
+    import imp
```
Comment (fabioz) [inline pydev_ipython/qt_loaders.py:130]: This seems backward... `importlib` should be preferred over `imp` which was deprecated (so, I think the version using `importlib` should be kept -- unless you see a real reason for the switch).

File: pydevd.py (hunk 7 above)
Comment (fabioz) [reply pydevd.py:1084]: Ok... this really needs some debugging (as I don't understand why `get_file_type` would return something that's not `PYDEV_FILE` for something in `sys_monitoring` -- if something is needed, the fix should probably be there as this would affect other places too, but in theory this shouldn't be needed, possibly something is different when running in debugpy).

File: pydevd_file_utils.py [hunk 25]
```diff
@@ -963,3 +963,15 @@ def get_package_dir(mod_name):
         if os.path.isdir(mod_path):
             return mod_path
     return None
+
+def contains_dir(path, dir_name):
```
Comment (fabioz) [inline pydevd_file_utils.py:967]: Is this used anywhere? Should it be removed?

File: tests_python/test_debugger.py [hunk 26]
```diff
@@ -3044,7 +3044,7 @@ def test_attach_to_pid_no_threads(case_setup_remote, reattach):
         writer.finished_ok = True
 
 
-@pytest.mark.skipif(not IS_CPYTHON or IS_MAC or not SUPPORT_ATTACH_TO_PID, reason="CPython only test (brittle on Mac).")
+@pytest.mark.skipif(True, reason="Test is flaky.")
```
Comment (fabioz) [inline tests_python/test_debugger.py:3047]: Humm, I think this wasn't failing before (on CPython with py 3.6 to 3.11 without mac). Isn't this the case anymore?

Comment (rchiodo) [reply tests_python/test_debugger.py:3047]: It's flakey. Fails on some runs. Failed on 3.11 ubuntu once, then didn't next run.

File: pydevd_file_utils.py (hunk 25 above)
Comment (rchiodo) [reply pydevd_file_utils.py:967]: No it can be removed.

File: pydev_ipython/qt_loaders.py (hunk 24 above)
Comment (rchiodo) [reply pydev_ipython/qt_loaders.py:130]: I might have merged this wrong. I'll switch it back.

File: _pydevd_bundle/pydevd_cython.pyx (hunk 21 above)
Comment (rchiodo) [reply _pydevd_bundle/pydevd_cython.pyx:1419]: Yeah this was from the change I made from django. This change isn't necessary anymore. I'll revert it.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 3 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:215]: Why is the cache needed in the first place? It seems like everytime an unwind event occurred it should be reset as it could be for an entirely different exception. At the very least, shouldn't the cache also keep track of the owning exception?

File: _pydevd_bundle/pydevd_process_net_command_json.py (hunk 22 above)
Comment (rchiodo) [reply _pydevd_bundle/pydevd_process_net_command_json.py:348]: Whoops, yeah missed that.

File: pydev_ipython/inputhookqt4.py (hunk 23 above)
Comment (rchiodo) [reply pydev_ipython/inputhookqt4.py:22]: I didn't make this change, it was in debugpy. I'll revert it.

File: pydevd.py (hunk 7 above)
Comment (rchiodo) [reply pydevd.py:1084]: Okay found the problem. In debugpy the tests call pydevd.settrace (pydevd tests don't do this). Internally this calls:

```
pydevd::_locked_settrace
pydevd::set_tracing_for_untraced_contexts
pydevd::set_trace_for_frame_and_parents line 2410
lambda
```

That lambda is causing a failure on the stop_on_entry tests for debugpy because it gets marked as the entry point.

With this change here, we skip tracing that lambda.

I'm working on adding a test case to pydevd for this.

Comment (rchiodo) [reply pydevd.py:1084]: Without Cython enabled, a similar situation happens with pydevd_sys_monitoring.py - a lambda gets marked as having file type 'none'. I have a better change I think though that will use the get_file_type instead of hardcoding it like is done here. I'll submit that once I figure out a test that reproduces the problem.

File: tests_python/test_debugger.py (hunk 26 above)
Comment (rchiodo) [reply tests_python/test_debugger.py]: Trying it again. Fails locally for me for 3.12 now

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 3 above)
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py]: I updated the cache to clear itself it the cached frame isn't in the callstack for the new frame (which I believe should mean that the old frame was from a different exception).

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 27]
```diff
@@ -177,15 +178,30 @@ def _get_unhandled_exception_frame(depth: int) -> Optional[FrameType]:
...
+            return None
+
+        f_unhandled = _getframe(depth)
+        search_depth = 0
 
         while f_unhandled is not None and f_unhandled.f_back is not None:
             f_back = f_unhandled.f_back
             filename = f_back.f_code.co_filename
             name = splitext(basename(filename))[0]
+            search_depth += 1
```
Comment (rchiodo) [inline _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:204]: Whoops, forgot to remove this.

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 3 above)
Comment (fabioz) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py]: Well, the cache is a performance improvement thing (so, for correctness we could remove it for now, but for performance it may be good to add it back in the future if possible).

A note: in the refactoring you called it `_is_user_frame`, but that's not really correct, what it's trying to discover here is handling for unhandled exceptions and it's trying to determine if this is the frame that when reached the exception has to be considered unhandled (meaning it's the last frame before actually being raised as unhandled) -- the thing is that (in theory) this shouldn't really change for the same thread after it's computed once (apart from the corner case you found where apparently this is being computed too soon).

File: _pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 28]
```diff
@@ -205,6 +219,17 @@ def _get_unhandled_exception_frame(depth: int) -> Optional[FrameType]:
...
 
+            elif name == "<frozen runpy>":
+                if f_back.f_code.co_name.startswith(("run", "_run")):
+                    break
+
+            elif name == "runpy":
+                if f_back.f_code.co_name.startswith(("run", "_run")):
+                    break
+            
+            elif py_db.get_file_type(f_back) is PYDEV_FILE:
```
[the remark this answers is not in the document]
Comment (rchiodo) [reply _pydevd_sys_monitoring/_pydevd_sys_monitoring.py:230]: I noticed a pattern and thought maybe this would catch other things that were missed in the if statements above. But I don't recall this being necessary. I'll remove it and see if everything still passes.

**Comment by fabioz** [comment]
Just merged. Thank you for the PR!

**Comment by rchiodo** [comment]
Thanks a lot for the feedback. I'll be submitting the changes for 3.13 next. Those were much smaller though :)
